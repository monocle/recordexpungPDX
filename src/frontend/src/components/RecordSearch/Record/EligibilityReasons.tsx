import React from "react";
import { convertCaseNumberIntoLinks, capitalize } from "./util";
import { TypeEligibilityData, TimeEligibilityData } from "./types";

type EligibilityKind = "type" | "time";

interface EligibilityReasonsProps {
  kind: EligibilityKind;
  eligibility: TypeEligibilityData | TimeEligibilityData;
}

function Reason({
  reason,
  isFirst,
  kind,
}: {
  reason: string;
  isFirst: boolean;
  kind: EligibilityKind;
}) {
  return (
    <div className={isFirst ? "" : "bt b--light-gray pt2 mt2"}>
      <span className="fw7">{isFirst ? `${capitalize(kind)}: ` : "OR "}</span>
      {convertCaseNumberIntoLinks(reason)}
    </div>
  );
}

const typeStatusMap = new Map([
  ["eligible", { icon: "circle", color: "gray" }],
  ["ineligible", { icon: "times-circle", color: "red" }],
  ["needs more analysis", { icon: "question-circle", color: "purple" }],
]);

const timeStatusMap = new Map([
  ["eligible", { icon: "check-circle", color: "green" }],
]);

export default function Reasons({
  kind,
  eligibility,
}: EligibilityReasonsProps) {
  const { status, reason: reasonString } = eligibility;
  const isEligibleNow = kind === "time" && status === "Eligible";
  const reasons = reasonString.split("OR");
  const statusMap = kind === "time" ? timeStatusMap : typeStatusMap;
  let props = statusMap.get(status.toLowerCase());

  if (!props) {
    if (kind === "type") {
      return <>Unknown type eligibility</>;
    }

    props = {
      icon: "clock",
      color: "dark-blue",
    };
  }

  return (
    <div className={"relative mb3 connect connect-" + kind}>
      <i
        aria-hidden="true"
        className={`absolute fas fa-${props.icon} ${props.color} bg-white outline-2-white z-1`}
      ></i>
      <div className="ml3 pl1">
        {isEligibleNow ? (
          <>
            <span className="fw7">Time:</span> Eligible Now
          </>
        ) : (
          reasons.map((reason: string, index: number) => {
            return (
              <Reason
                key={reason}
                reason={reason}
                isFirst={index === 0}
                kind={kind}
              />
            );
          })
        )}
      </div>
    </div>
  );
}
