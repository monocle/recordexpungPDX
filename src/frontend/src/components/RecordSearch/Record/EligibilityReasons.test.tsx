import React from "react";
import "@testing-library/jest-dom";
import renderer from "react-test-renderer";
import { TimeEligibilityData, TypeEligibilityData } from "./types";
import EligibilityReasons from "./EligibilityReasons";

const timeEligibility = {
  status: "",
  reason: "a good reason",
  date_will_be_eligible: "a long time",
};

function doRenderTime(time_eligibility: TimeEligibilityData) {
  return renderer
    .create(<EligibilityReasons kind="time" eligibility={time_eligibility} />)
    .toJSON();
}

describe("For time eligibility", () => {
  it("renders correctly for an 'Eligible' status", () => {
    const time_eligibility = { ...timeEligibility, status: "Eligible" };
    expect(doRenderTime(time_eligibility)).toMatchSnapshot();
  });

  it("renders correctly for a non specific status", () => {
    const time_eligibility = {
      ...timeEligibility,
      status: "not a real status",
    };
    expect(doRenderTime(time_eligibility)).toMatchSnapshot();
  });
});

const typeEligibility = {
  status: "",
  reason: "a good reason",
};

function doRenderType(type_eligibility: TypeEligibilityData) {
  return renderer
    .create(<EligibilityReasons kind="type" eligibility={type_eligibility} />)
    .toJSON();
}

describe("For type eligibility", () => {
  it("renders correctly for an 'Eligible' status", () => {
    const type_eligibility = { ...typeEligibility, status: "Eligible" };
    expect(doRenderType(type_eligibility)).toMatchSnapshot();
  });

  it("renders correctly for an 'Ineligible' status", () => {
    const type_eligibility = { ...typeEligibility, status: "Ineligible" };
    expect(doRenderType(type_eligibility)).toMatchSnapshot();
  });

  it("renders correctly for an 'Needs More Analysis' status", () => {
    const type_eligibility = {
      ...typeEligibility,
      status: "Needs More Analysis",
    };
    expect(doRenderType(type_eligibility)).toMatchSnapshot();
  });

  it("renders correctly for a non specific status", () => {
    const type_eligibility = {
      ...typeEligibility,
      status: "not a real status",
    };
    expect(doRenderType(type_eligibility)).toMatchSnapshot();
  });
});
