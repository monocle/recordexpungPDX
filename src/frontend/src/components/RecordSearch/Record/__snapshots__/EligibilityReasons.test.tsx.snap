// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`For time eligibility renders correctly for a non specific status 1`] = `
<div
  className="relative mb3 connect connect-time"
>
  <i
    aria-hidden="true"
    className="absolute fas fa-clock dark-blue bg-white outline-2-white z-1"
  />
  <div
    className="ml3 pl1"
  >
    <div
      className=""
    >
      <span
        className="fw7"
      >
        Time: 
      </span>
      a good reason
    </div>
  </div>
</div>
`;

exports[`For time eligibility renders correctly for an 'Eligible' status 1`] = `
<div
  className="relative mb3 connect connect-time"
>
  <i
    aria-hidden="true"
    className="absolute fas fa-check-circle green bg-white outline-2-white z-1"
  />
  <div
    className="ml3 pl1"
  >
    <span
      className="fw7"
    >
      Time:
    </span>
     Eligible Now
  </div>
</div>
`;

exports[`For type eligibility renders correctly for a non specific status 1`] = `"Unknown type eligibility"`;

exports[`For type eligibility renders correctly for an 'Eligible' status 1`] = `
<div
  className="relative mb3 connect connect-type"
>
  <i
    aria-hidden="true"
    className="absolute fas fa-circle gray bg-white outline-2-white z-1"
  />
  <div
    className="ml3 pl1"
  >
    <div
      className=""
    >
      <span
        className="fw7"
      >
        Type: 
      </span>
      a good reason
    </div>
  </div>
</div>
`;

exports[`For type eligibility renders correctly for an 'Ineligible' status 1`] = `
<div
  className="relative mb3 connect connect-type"
>
  <i
    aria-hidden="true"
    className="absolute fas fa-times-circle red bg-white outline-2-white z-1"
  />
  <div
    className="ml3 pl1"
  >
    <div
      className=""
    >
      <span
        className="fw7"
      >
        Type: 
      </span>
      a good reason
    </div>
  </div>
</div>
`;

exports[`For type eligibility renders correctly for an 'Needs More Analysis' status 1`] = `
<div
  className="relative mb3 connect connect-type"
>
  <i
    aria-hidden="true"
    className="absolute fas fa-question-circle purple bg-white outline-2-white z-1"
  />
  <div
    className="ml3 pl1"
  >
    <div
      className=""
    >
      <span
        className="fw7"
      >
        Type: 
      </span>
      a good reason
    </div>
  </div>
</div>
`;
